custom-targets/vertex-ai/model-deployer/deploy.go

```diff
@@ -0,0 +1,304 @@
+// Copyright 2023 Google LLC
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     https://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"google.golang.org/api/aiplatform/v1"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/GoogleCloudPlatform/cloud-deploy-samples/custom-targets/util/clouddeploy"
+	"sigs.k8s.io/yaml"
+
+	"cloud.google.com/go/storage"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+const aiDeployerSampleName = "clouddeploy-vertex-ai-sample"
+
+// deployer implements the handler interface to deploy a model using the vertex AI API.`.
+type deployer struct {
+	gcsClient *storage.Client
+	params    *params
+	req       *clouddeploy.DeployRequest
+}
+
+// process processes the Deploy request.
+func (d *deployer) process(ctx context.Context) error {
+	fmt.Println("Processing deploy request")
+
+	res, err := d.deploy(ctx)
+	if err != nil {
+		fmt.Printf("Deploy failed: %v\n", err)
+		dr := &clouddeploy.DeployResult{
+			ResultStatus:   clouddeploy.DeployFailed,
+			FailureMessage: err.Error(),
+		}
+		d.addCommonMetadata(res)
+		fmt.Println("Uploading failed deploy results")
+		rURI, err := d.req.UploadResult(ctx, d.gcsClient, dr)
+		if err != nil {
+			return fmt.Errorf("error uploading failed deploy results: %v", err)
+		}
+		fmt.Printf("Uploaded failed deploy results to %s\n", rURI)
+		return err
+	}
+	d.addCommonMetadata(res)
+
+	fmt.Println("Uploading successful deploy results")
+	rURI, err := d.req.UploadResult(ctx, d.gcsClient, res)
+	if err != nil {
+		return fmt.Errorf("error uploading deploy results: %v", err)
+	}
+	fmt.Printf("Uploaded deploy results to %s\n", rURI)
+	return nil
+
+}
+
+func (d *deployer) deploy(ctx context.Context) (*clouddeploy.DeployResult, error) {
+	localManifest := "manifest.yaml"
+	fmt.Printf("Downloading deploy input manifest from %q.\n", d.req.ManifestGCSPath)
+
+	_, err := d.req.DownloadManifest(ctx, d.gcsClient, localManifest)
```

Comment:
I advocate for logging where the manifest was downloaded from so the user can more easily debug from the build logs if necessary.